docs/source/federation/gateway.mdx

```diff
@@ -0,0 +1,227 @@
+});
+
+server.listen().then(({ url }) => {
+  console.log(`🚀 Server ready at ${url}`);
+});
+```
+
+In the above example, we provide the `serviceList` configuration option to the
+`ApolloGateway` constructor. This array specifies a `name` and `url` for each
+of our implementing services. You can specify any string value for `name`, which
+is used primarily for query planner output, error messages, and logging.
+
+> In production, we recommend running the gateway in a **managed mode**, which relies on static files rather than introspection. For details, see the [Apollo Graph Manager documentation](https://www.apollographql.com/docs/graph-manager/federation/).
+
+On startup, the gateway fetches each implementing service's schema from its `url` and composes those schemas into a single federated data graph. It then begins accepting incoming requests and creates query plans for them that execute across one or more services.
+
+> If there are any composition errors, the `new ApolloServer` call throws an exception
+> with a list of [validation errors](/federation/errors/).
+
+## Customizing requests and responses
+
+The gateway can modify the details of an incoming request before executing it across your implementing services. For example, your services might all use the same authorization token to associate an incoming request with a particular user. The gateway can add that token to each operation it sends to your services.
+
+Similarly, the gateway can modify the details of its response to a client, based on the result returned by each implementing service.
+
+### Customizing requests
+
+In the following example, each incoming request to the gateway includes an `Authorization` header. The gateway sets the shared `context` for an operation  by pulling the value of that header and using it to fetch the associated user's ID.
+
+After adding the `userId` to the shared `context` object, the gateway can then add that value to a header that it includes in its requests to each implementing service.
+
+<ExpansionPanel title="Expand example">
+
+```javascript
+const { ApolloServer } = require('apollo-server');
+const { ApolloGateway, RemoteGraphQLDataSource } = require('@apollo/gateway');
+
+// highlight-start
+class AuthenticatedDataSource extends RemoteGraphQLDataSource {
+  willSendRequest({ request, context }) {
+    // pass the user's id from the context to underlying services
+    // as a header called `user-id`
+    request.http.headers.set('user-id', context.userId);
+  }
+}
+// highlight-end
+
+const gateway = new ApolloGateway({
+  serviceList: [
+    { name: 'products', url: 'http://localhost:4001' },
+    // other services
+  ],
+  // highlight-start
+  buildService({ name, url }) {
+    return new AuthenticatedDataSource({ url });
+  },
+  // highlight-end
+});
+
+const server = new ApolloServer({
+  gateway,
+
+  // Disable subscriptions (not currently supported with ApolloGateway)
+  subscriptions: false,
+
+  context: ({ req }) => {
```

Comment:
I know this focuses on `apollo-server`, which uses `apollo-server-express` (and eventually, `express`) under the the hood, but in the same spirit as my comment on https://github.com/apollographql/apollo-server/pull/3834/files#r391677331, it may be worth a non-normative note that specifies that this will be different on non-`express` integrations (e.g. Fastify, Koa, AWS Lambda, Google Cloud Functions).

The risk of course is copy-pasta and subtly unexpected behavior that results from such, but it's a state we've seen users encounter a few times and is slated to be improved in the implementation, _eventually_.